biggus/__init__.py

```diff
@@ -936,6 +936,131 @@ def _groups_of(length, total_length):
     return _pairwise(indices)
 
 
+class TransposedArray(Array):
+    def __init__(self, array, axes=None):
+        """
+        Permute the dimensions of an array.
+
+        Parameters
+        ----------
+        array - array like
+            The array to transpose
+        axes - list of ints, optional
+            By default, reverse the dimensions, otherwise permute the axes
+            according to the values given.
+
+        """
+        self.pre_transposed = array
+        if axes is None:
+            axes = np.arange(array.ndim)[::-1]
+        elif len(axes) != array.ndim:
+            raise ValueError('Incorrect number of dimensions.')
+        self.axes = axes
+        self._forward_axes_map = {src: dest for dest, src in enumerate(axes)}
+        self._inverse_axes_map = {dest: src for dest, src in enumerate(axes)}
+
+    def __repr__(self):
+        return 'TransposedArray({!r}, {!r})'.format(self.pre_transposed,
+                                                    self.axes)
+
+    def _apply_axes_mapping(self, target, inverse=False):
+        """
+        Apply the transposition to the target iterable.
+
+        Parameters
+        ----------
+        target - iterable
+            The iterable to transpose. This would be suitable for things
+            such as a shape as well as a list of ``__getitem__`` keys.
+        inverse - bool
+            Whether to map old dimension to new dimension (forward), or
+            new dimension to old dimension (inverse). Default is False
+            (forward).
+
+        Returns
+        -------
+        A tuple derived from target which has been ordered based on the new
+        axes.
+
+        """
+        if len(target) != self.ndim:
+            raise ValueError('The target iterable is of length {}, but '
+                             'should be of length {}.'.format(len(target),
+                                                              self.ndim))
+        if inverse:
+            axis_map = self._inverse_axes_map
+        else:
+            axis_map = self._forward_axes_map
+
+        result = [None] * self.ndim
+        for axis, item in enumerate(target):
+            result[axis_map[axis]] = item
+        return tuple(result)
+
+    @property
+    def ndim(self):
+        return self.pre_transposed.ndim
+
+    @property
+    def dtype(self):
+        return self.pre_transposed.dtype
+
+    @property
+    def shape(self):
+        return self._apply_axes_mapping(self.pre_transposed.shape)
+
+    def __getitem__(self, keys):
+        keys = _full_keys(keys, self.ndim)
+        new_transpose_order = self.axes[:]
+
+        # Split the keys into np.newaxis and normal keys.
+        new_axes = [axis for axis, key in enumerate(keys) if key is None]
```

Comment:
I'm sure this didn't pass you by @rhattersley - TransposedArrays can be indexed with ``np.newaxis``. Essentially, if you were desperate to index with newaxis, is *would* be possible to do:

```
TransposedArray(lazy_array, range(lazy_array.ndim))[..., np.newaxis, :, np.newaxis]
```

I'm not saying that is a thing you'd want to do, but it is a fallback option (and hopefully shows that ``_full_keys`` is helpful for interpreting the keys passed to ``__getitem__``).